### ISSUE TITLE
MRView tractography tool: track files not loading properly

### ISSUE THREAD

**Issue by jdtournier** [MRtrix3/mrtrix3#278]
If I load a track file using the normal open tracks dialog, the streamlines come out in very funny places (along the edges of the FOV). Things work fine when loading the same tracks using the `-tractography.load` command-line option. Things work fine on the version prior to the merge with the QOpenGLWidget branch. This issue is specific to Qt >= 5.4 - no problem with Qt 4.8. This is a bit weird, since none of the changes introduced with that branch had anything to do with tractography... I think this might be related to #271 - there might be a subtle OpenGL bug in there that only shows up under specific circumstances...

**Comment by thijsdhollander** [comment]
For the record: I'm not having this issue here (Qt 5.2.1, GeForce GTX 650, 3.3.0 NVIDIA 304.125 drivers).  Tracks are looking correct, and come out the same whether loaded through the dialog or the command-line option.

**Comment by rtabbara** [comment]
@jdtournier I was unable to replicate this or #271 on my machine, however I have attempted to tighten things up with some changes that I've put into the branch <code>tractogram_fixup</code>. I've also improved upon my initial naive attempt at downsampling, which currently is adjusted by line thickness.  Please check it out.

**Comment by jdtournier** [comment]
OK, as I'd mentioned above, this issue seems to be specific to the changes introduced for Qt 5.4 - these don't seem to apply to earlier versions (with exceptions, see #271, and below). If you're running Ubuntu 14.04, you'll be using Qt 5.2, which still uses the previous API. So not too surprised you'd be having trouble replicating...

Anyway, I gave your `tractogram_fixup` branch a whirl, and it runs smooth on Qt 4.8. No go on Qt 5.4 though, here's a screenshot:
![screenshot from 2015-06-24 09-29-46](https://cloud.githubusercontent.com/assets/3221000/8325858/8a772eac-1a54-11e5-8130-1f71f587eeda.png)
Note that the image in the background has disappeared - I've not windowed it to black. Even after closing the tractogram and tractography tool, I don't get the image back. Seems to me like some kind of memory access issue, either on the CPU or the GPU (most likely the latter), messing up some internal structure somewhere.

Note that this only happens when I load the tracks from the tractography tool - it works fine if I load the tractogram from the command-line. 

Needless to say, things work fine on a Qt 4.8 build. 

Here's a couple of screenshots of the same data opened with the master branch version, to illustrate some of my earlier comments:
![screenshot from 2015-06-24 09-31-22](https://cloud.githubusercontent.com/assets/3221000/8326124/b38e3928-1a56-11e5-9121-1dc93523a1da.png)
![screenshot from 2015-06-24 09-31-48](https://cloud.githubusercontent.com/assets/3221000/8326127/b686dd06-1a56-11e5-9173-6831b5d654ed.png)

And finally, here's one on a small ROI (from the testing repo), to show the issue with the black ends I mentioned before:
![screenshot from 2015-06-24 09-50-41](https://cloud.githubusercontent.com/assets/3221000/8326176/17908ade-1a57-11e5-960a-29cec404ee98.png)

So there's something going on, but I'm not sure what's causing it. I've had a close look at the changes I made to support Qt 5.4's QOpenGLWidget tool, but I can't think what I could have done to cause this - there's no memory allocation of any kind, the only thing that changes the GL state are a couple of calls to gl::ClearColor(), gl::ColorMask(), and gl::Clear() - none of which seem particularly likely to cause issues. But given that similar issues are also present on ATI hardware using Ubuntu 14.04 with Qt 4.8 (#271), I'm more inclined to think there is a subtle bug in the tractography tool, something like a buffer underrun or overrun on one of the vertex arrays during one of the gl::Draw calls, with consequences that would manifest differently depending on exactly how the context has been set up, which would influence what was being overrun, and also whether the driver does anything about it, etc. So I reckon the issue might very well be present in all situations, but only manifest as an overt bug under specific conditions...

**Comment by thijsdhollander** [comment]
Okay, I'm on to something... I've got a scenario where I can reliably produce a *real* variant of the true "black segment" issue.  And it's not necessarily "black" folks!  Let me see if I can get some facts straight, and then I'll enlist some of them...

**Comment by thijsdhollander** [comment]
Good, so, some raw facts:

1) I'm not running ATI hardware nor Qt 5.4, but I'm getting a true variant of the "black segments". (Nvidia hardware, Ubuntu 14.04, Qt 4.8)
2) In my specific case, I'm getting purely "black segments" at the end of some streamlines, specifically *only* at the bottom of the corticospinal tract.  But it's not `directly` related to z-coordinate or something, because they do appear for some "longer" as well as some "shorter" tracks down there, but then again also not for others: no obvious strict z-coordinate separation possible for the case I'm seeing.
3) I'm getting them in `master` as well as in the current state (as I'm writing this) of `tractogram_fixup`.
4) I'm getting them by opening the tractogram through the dialog in the GUI, as well as through the `-tractography.load` command line option.
5) Despite points (3) and (4), I seem to be getting a smaller number of black ends in `tractogram_fixup` relative to `master`, and also a smaller number of black ends by using the command line option relative to using the GUI dialog.  Using the new branch combined with the command line option was thus the "best", but I could still spot at least one clearly black segment in a 25000 track tractogram.
6) It's not suddenly happening because of a large number of tracks, but is of course more likely simply because there's more opportunities for this bug to manifest in that case.  Found this out the hard way, because I was trying to bulldoze mrview with 10 million tracks, not cropped to slab and with lighting on.  It rendered, with the black ends, and then basically froze my computer while still heating up the hardware up to the point I wish I'd put bacon and eggs on it for a good meal.  After a hard reboot, I tried with less tracks (first 1 million, and then just 25000).  The black ends still showed; I had the feeling it was a similar fraction of the total amount of tracks that were affected -- just a lower absolute number because of a lower absolute total number of tracks.  Just throwing more tracks at it, simply increases chances of spotting the effect somewhere... I guess.
7) It's not the background colour: setting that one to a vibrant green, I still saw *black* segments, not green ones.
8) My main image does not disappear.  Silly suggestion, and I'm sure you didn't do it, but still...: you're sure you didn't accidentally hit the `m` key or something (that's the current shortcut to hide the main image, also resulting in no apparent feedback from any windowing attempts...)?
9) And now for the really interesting part (I believe): it's not "just black".  It's black when using the `colour by direction` mode.  It's still the colour of the rest of the track when using either `colour by track ends`, `randomise colour` or `set colour` modes.  But, in even in these latter modes, when lighting is switched on, the lighting doesn't show for these specific "bugged" segments.  So going from raw facts to suggestions here: I get the strong feeling it relates to the ability to get specific information of these bugged segments... as maybe in: their orientation... which is needed for the colour when using `colour by direction`, and probably always for the lighting?
10) I *really* like the improvements @rtabbara has made in this new branch for the line thickness rescaling when zooming in or out.  It behaves so much better now, and I don't get slightly overlapping segments in some cases when making things partially transparent.  Also, lighting seems to "propagate" more continuous from segment to segment because of that.  It's definitely looking much sleeker in a wide range of scenarios now, and is actually still surprisingly performant.  I wasn't expecting much when throwing 1 million tracks at it, but I actually still get a framerate of 2 frames/sec... where I thought interaction would basically become impossible.  I'm quite impressed by this.
11) ...but playing around with the line thickness slider, I think I stumbled on a final important hint:  increasing line thickness (resulting in less samples), also seems to lower the amount of bugged endpoint segments.  So it's related to where/how is sampled (I think), running into issues in some (probably exceptional) cases at endpoints.
12) Playing around so much with things, I suddenly ended up crashing the viewer with a segfault upon switching that colouring mode a final time. :sweat_smile:  Focussing just on that and switching colour modes more often, I was able to trigger more segfaults. :fearful: 

So I have the feeling the answer to (at least) the black segments issue is related to points (9) and (11).  Otherwise, I'm happy to have played around with some awesome looking tracks. :smile: 

Here's a new screens:

Colour by direction:
![1](https://cloud.githubusercontent.com/assets/11952694/8347619/52b866aa-1b4b-11e5-8706-6c4dbacc3693.png)

Colour by track ends:
![2](https://cloud.githubusercontent.com/assets/11952694/8347621/5974ae22-1b4b-11e5-87af-0737e027d7c5.png)

Coloured by manually selecting a shade of red"
![3](https://cloud.githubusercontent.com/assets/11952694/8347626/67cb0642-1b4b-11e5-886b-5fb2738076d5.png)

Changed the lighting settings for these images, so the contrast between shaded and not-shaded is more clear in the latter ones.

**Comment by jdtournier** [comment]
Ok, thanks to @thijsdhollander for a _very_ thorough report. I think this probably supports what I was suggesting earlier: there's probably a buffer overrun at the ends of the tracks - or certainly at the ends of each vertex buffer (which might contain a number of tracks, if I recall correctly). That would explain the failure to get the correct tangent for colouring and lighting, and potentially also why some of the track segments go funny places in some of my cases. Might also explain the segfaults, although that one is harder to figure out - if this is happening on the GPU, it's hard to see how it could trigger a segfault on the GPU, although clearly this isn't impossible.

I guess the first thing to check is that the number of vertices supplied to the geometry shader genuinely matches up with the size of the vertex buffer, and that all the offsets, strides, and index buffers are appropriate. 

The other thing to check is that the NaN triplets at each end of each track are handled appropriately - I don't know how the different bits of GPU hardware might handle these in the computations. This one is unlikely though, these ought to conform to the IEEE floating-point standard (?).

One last thing: those interested in OpenGL debugging might want to look into [BuGLe](https://www.opengl.org/sdk/tools/BuGLe/) - it's a really neat way to pick up subtle errors, although a bit fiddly to set up. Happy to share my config if anyone is interested...

**Comment by rtabbara** [comment]
@thijsdhollander Thanks for taking a look at this. My guess is there is indeed a buffer overrun that is causing the artefacts you're observing.

@jdtournier The bad news is that I finally got my hands on Qt5.4 and I don't think this is the sole issue. In particular, using the vector plot tool I was also able to replicate the issue of a disappearing image. When I initially load the the vector file, both the vector and underlying image are visible. It's only when I attempt to adjust the current slice that the image disappears. What's stranger is that nothing is rendered (neither image or vector plot) when turning off slab cropping.

I was also able to replicate the tractogram artefacts you observed, but these seem to disappear/reappear when adjusting the field of view. Currently, my downsampling stride is not updated for a FOV change.

Anyway, need to investigate further...

**Comment by thijsdhollander** [comment]
Just for the sake of completeness, as I already told @rtabbara on the way to some good Mexican food yesterday: the segfaults I mentioned earlier have only occurred to me in (several) cases where I specifically switched from anything *but* the `colour by direction` mode, to the actual `colour by direction` mode.  I can't produce this bug consistently, but just switching around a lot while adjusting some other settings here and there, seems to eventually trigger it upon making *that specific switch* of colouring modes.  I haven't had a segfault happening (yet) upon switching lighting on.

**Comment by jdtournier** [comment]
OK, I finally gave [BuGLe](https://www.opengl.org/sdk/tools/BuGLe/) a spin, there's tons of errors showing up before I've even loaded an image... Tons more when I do load an image - and I haven't even bothered opening any tools yet. I think I'd better make sure things run clean for the main window before we worry about what's going on with the tractography tool...

**Comment by rtabbara** [comment]
@jdtournier Think I finally tracked this down. Please try out the latest from <code>tractogram_fixup</code>. 

The issue was that for the vector/tractogram tools the VAO's were being setup outside of <code>paintGL</code> and so we need to explicitly call <code>makeCurrent</code> to ensure that the correct graphics context is set prior to loading up our vertex buffers.

My guess is that this issue pops up in Qt5.4 because <code>QOpenGLWidget</code> has its own gl context that is distinct from the window.

**Comment by jdtournier** [comment]
@rtabbara Good get! This does indeed fix up most of the problems - at least on my laptop (I'll test it out on my desktop and the other ATI desktop tomorrow). However, this didn't sort out the issues with the black track ends, unfortunately. I think there's still a buffer overrun or something in there, but it certainly works a lot better now...

**Comment by jdtournier** [comment]
@rtabbara : you'll notice I've just merged your branch to master - basically master was unusable for me without it, and it seemed to work fine for both qt4 & qt5, so I just went ahead. We'll still need to figure out the black ends business though...

**Comment by thijsdhollander** [comment]
Yep, I can confirm here: I still have the "black ends" as well (never had the other issues, those were probably indeed specific to Qt 5.4).  I checked yesterday evening, right before and right after I pulled the commit, and the "black ends" seem to appear in the exact same locations as they did before (so it at least seems they weren't affected in any way by the changes).  As I mentioned before, I'm also very happy with the changes to the downsampling strategy; good thing they're already in master now! :-)
